File: src/main/java/ca/rafemurray/BoosterBlock.java
package ca.rafemurray;

import com.mojang.serialization.MapCodec;
import net.minecraft.block.*;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityCollisionHandler;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.BooleanProperty;
import net.minecraft.state.property.EnumProperty;
import net.minecraft.state.property.Properties;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.Vec3d;
import net.minecraft.util.math.random.Random;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.util.shape.VoxelShapes;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;
import net.minecraft.world.WorldView;
import net.minecraft.world.tick.ScheduledTickView;

//TODO: Add textures for "connected" boosters
public class BoosterBlock extends ConnectingBlock{
    public static final MapCodec<BoosterBlock> CODEC = createCodec(BoosterBlock::new);
    public static final EnumProperty<Direction> FACING = Properties.FACING;
    public static final BooleanProperty NORTH = ConnectingBlock.NORTH;
    public static final BooleanProperty EAST = ConnectingBlock.EAST;
    public static final BooleanProperty SOUTH = ConnectingBlock.SOUTH;
    public static final BooleanProperty WEST = ConnectingBlock.WEST;
    public static final BooleanProperty UP = ConnectingBlock.UP;
    public static final BooleanProperty DOWN = ConnectingBlock.DOWN;
    public BoosterBlock(AbstractBlock.Settings settings) {
        super(16.0f, settings);
        setDefaultState(getDefaultState()
                .with(NORTH, false)
                .with(SOUTH, false)
                .with(EAST, false)
                .with(WEST, false)
                .with(UP, false)
                .with(DOWN, false)
                .with(FACING, Direction.NORTH));
    }

    @Override
    protected boolean isTransparent(BlockState state) {
        return true;
    }

    protected VoxelShape getCameraCollisionShape(BlockState state, BlockView world, BlockPos pos, ShapeContext context) {
        return VoxelShapes.empty();
    }

    protected float getAmbientOcclusionLightLevel(BlockState state, BlockView world, BlockPos pos) {
        return 1.0F;
    }

    protected boolean isSideInvisible(BlockState state, BlockState stateFrom, Direction direction) {
        return stateFrom.isOf(this) || super.isSideInvisible(state, stateFrom, direction);
    }

    public BlockState getPlacementState(ItemPlacementContext ctx) {
        BlockView world = ctx.getWorld();
        BlockPos pos = ctx.getBlockPos();
        BlockPos northPos = pos.north();
        BlockPos southPos = pos.south();
        BlockPos eastPos = pos.east();
        BlockPos westPos = pos.west();
        BlockPos upPos = pos.up();
        BlockPos downPos = pos.down();
        BlockState north = world.getBlockState(northPos);
        BlockState south = world.getBlockState(southPos);
        BlockState east = world.getBlockState(eastPos);
        BlockState west = world.getBlockState(westPos);
        BlockState up = world.getBlockState(upPos);
        BlockState down = world.getBlockState(downPos);
        Direction facing = ctx.getPlayerLookDirection();
        return (BlockState)this.getDefaultState()
                .with(FACING, facing)
                .with(NORTH, canConnect(north, facing))
                .with(SOUTH, canConnect(south, facing))
                .with(EAST, canConnect(east, facing))
                .with(WEST, canConnect(west, facing))
                .with(UP, canConnect(up, facing))
                .with(DOWN, canConnect(down, facing));
    }

    private boolean canConnect(BlockState neighbour, Direction facing) {
        if (!(neighbour.getBlock() instanceof BoosterBlock)) {
            return false;
        }
        if (neighbour.get(FACING) != facing) {
            return false;
        }
        return true;
    }

    protected void onEntityCollision(BlockState state, World world, BlockPos pos, Entity entity, EntityCollisionHandler handler) {
        if (entity instanceof PlayerEntity) {
            this.boost(state, entity);
        }
    }

    private void boost(BlockState state, Entity entity) {
        Vec3d velocity = entity.getVelocity();
        if (!state.contains(FACING)) return;
        Direction facing = state.get(FACING);

        double x = velocity.x;
        double y = velocity.y;
        double z = velocity.z;
        switch (facing) {
            case NORTH -> z -= 0.5;
            case SOUTH -> z += 0.5;
            case EAST -> x += 0.5;
            case WEST -> x -= 0.5;
            case UP -> y += 0.25;
            case DOWN -> y -= 0.25;
        }
        entity.setVelocity(x,y,z);
    }

    @Override
    protected BlockState getStateForNeighborUpdate(BlockState state, WorldView world, ScheduledTickView tickView, BlockPos pos, Direction direction, BlockPos neighborPos, BlockState neighborState, Random random) {
        return state.with(FACING_PROPERTIES.get(direction), canConnect(neighborState, state.get(FACING)));
    }

    @Override
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
        builder.add(FACING, NORTH, SOUTH, EAST, WEST, UP, DOWN);
    }

    @Override
    protected MapCodec<BoosterBlock> getCodec() {
        return CODEC;
    }
}

File: src/main/java/ca/rafemurray/block/BoosterBlock.kt
package ca.rafemurray.block

import com.mojang.serialization.MapCodec
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.util.RandomSource
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.InsideBlockEffectApplier
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.Level
import net.minecraft.world.level.LevelReader
import net.minecraft.world.level.ScheduledTickAccess
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.PipeBlock
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.StateDefinition
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import net.minecraft.world.level.block.state.properties.EnumProperty
import net.minecraft.world.level.block.state.properties.Property
import net.minecraft.world.phys.shapes.CollisionContext
import net.minecraft.world.phys.shapes.Shapes
import net.minecraft.world.phys.shapes.VoxelShape

//TODO: Add textures for "connected" boosters
class BoosterBlock(properties: Properties) : PipeBlock(16.0f, properties) {
    init {
        registerDefaultState(
            stateDefinition.any()
                .setValue(NORTH, false)
                .setValue(SOUTH, false)
                .setValue(EAST, false)
                .setValue(WEST, false)
                .setValue(UP, false)
                .setValue(DOWN, false)
                .setValue(FACING, Direction.NORTH)
        )
    }

    override fun entityInside(
        state: BlockState, level: Level, pos: BlockPos,
        entity: Entity, effectApplier: InsideBlockEffectApplier,
        bl: Boolean
    ) {
        if (entity is Player) {
            this.boost(state, entity)
        }
    }

    private fun boost(state: BlockState, entity: Entity) {
        val velocity = entity.deltaMovement
        if (!state.`is`(ModBlocks.BOOSTER)) return
        val facing = state.getValue(FACING)

        var x = velocity.x
        var y = velocity.y
        var z = velocity.z
        when (facing) {
            Direction.NORTH -> z -= 0.5
            Direction.SOUTH -> z += 0.5
            Direction.EAST -> x += 0.5
            Direction.WEST -> x -= 0.5
            Direction.UP -> y += 0.25
            Direction.DOWN -> y -= 0.25
            null -> {}
        }
        entity.setDeltaMovement(x, y, z)
    }

    override fun getStateForPlacement(context: BlockPlaceContext): BlockState {
        return getStateWithConnections(
            context.level, context.clickedPos, context.clickedFace,
            this.defaultBlockState()
        )
    }

    override fun updateShape(
        blockState: BlockState,
        levelReader: LevelReader,
        scheduledTickAccess: ScheduledTickAccess,
        blockPos: BlockPos,
        direction: Direction,
        blockPos2: BlockPos,
        other: BlockState,
        randomSource: RandomSource
    ): BlockState {
        val canConnect = other.`is`(this) && canConnect(other, blockState.getValue(FACING))
        return blockState.setValue(PROPERTY_BY_DIRECTION[direction] as Property<Boolean>, canConnect)
    }

    override fun createBlockStateDefinition(builder: StateDefinition.Builder<Block, BlockState>) {
        builder.add(FACING, NORTH, EAST, SOUTH, WEST, UP, DOWN)
    }

    override fun codec(): MapCodec<BoosterBlock> {
        return CODEC
    }

    override fun getVisualShape(
        blockState: BlockState,
        blockGetter: BlockGetter,
        blockPos: BlockPos,
        collisionContext: CollisionContext
    ): VoxelShape {
        return Shapes.empty()
    }

    override fun getShadeBrightness(blockState: BlockState, blockGetter: BlockGetter, blockPos: BlockPos): Float {
        return 1.0f
    }

    override fun propagatesSkylightDown(blockState: BlockState): Boolean {
        return true
    }

    companion object {
        val CODEC: MapCodec<BoosterBlock> = simpleCodec { properties: Properties ->
            BoosterBlock(
                properties
            )
        }
        val FACING: EnumProperty<Direction> = BlockStateProperties.FACING

        private fun canConnect(neighbour: BlockState, facing: Direction): Boolean {
            if (neighbour.block !is BoosterBlock) {
                return false
            }
            if (neighbour.getValue(FACING) != facing) {
                return false
            }
            return true
        }

        fun getStateWithConnections(
            blockGetter: BlockGetter,
            blockPos: BlockPos,
            direction: Direction,
            blockState: BlockState
        ): BlockState {
            val below = blockGetter.getBlockState(blockPos.below())
            val above = blockGetter.getBlockState(blockPos.above())
            val north = blockGetter.getBlockState(blockPos.north())
            val south = blockGetter.getBlockState(blockPos.south())
            val east = blockGetter.getBlockState(blockPos.east())
            val west = blockGetter.getBlockState(blockPos.west())
            return blockState
                .setValue(DOWN, canConnect(below, direction))
                .setValue(UP, canConnect(above, direction))
                .setValue(NORTH, canConnect(north, direction))
                .setValue(SOUTH, canConnect(south, direction))
                .setValue(EAST, canConnect(east, direction))
                .setValue(WEST, canConnect(west, direction))
                .setValue(FACING, direction)
        }
    }
}
